pushsync/pusher.go

```diff
@@ -0,0 +1,333 @@
+				break
+			}
+			// increment synced count for the tag if exists
+			tag := item.tag
+			if tag != nil {
+				tag.Inc(chunk.StateSynced)
+				if tag.Done(chunk.StateSynced) {
+					p.logger.Debug("closing root span for tag", "taguid", tag.Uid, "tagname", tag.Name)
+					tag.FinishRootSpan()
+				}
+				// finish span for pushsync roundtrip, only have this span if we have a tag
+				item.span.Finish()
+			}
+
+			totalDuration := time.Since(item.sentAt)
+			metrics.GetOrRegisterResettingTimer("pusher.chunk.roundtrip", nil).Update(totalDuration)
+			metrics.GetOrRegisterCounter("pusher.receipts.synced", nil).Inc(1)
+
+			// collect synced addresses and corresponding items to do subsequent batch operations
+			syncedAddrs = append(syncedAddrs, addr)
+			item.synced = true
+
+			// retry interval timer triggers starting from new
+		case <-timer.C:
+			// initially timer is set to go off as well as every time we hit the end of push index
+			// so no wait for retryInterval needed to set  items synced
+			func() {
+				defer metrics.GetOrRegisterResettingTimer("pusher.mark-and-sweep", nil).UpdateSince(time.Now())
+
+				// if subscribe was running, stop it
+				if unsubscribe != nil {
+					unsubscribe()
+				}
+
+				// delete from pushed items
+				for i := 0; i < len(syncedAddrs); i++ {
+					delete(p.pushed, syncedAddrs[i].Hex())
+				}
+
+				// set chunk status to synced, insert to db GC index
+				if err := p.store.Set(ctx, chunk.ModeSetSync, syncedAddrs...); err != nil {
+					log.Error("pushsync: error setting chunks to synced", "err", err)
+				}
+
+				// reset synced list
+				syncedAddrs = nil
+
+				// we don't want to record the first iteration
+				if chunksInBatch != -1 {
+					// hack: this measurement is NOT a timer, but we want a histogram for chunks in batch, so it fits the data structure
+					metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.hist", nil).Update(time.Duration(chunksInBatch))
+
+					metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.time", nil).UpdateSince(batchStartTime)
+					metrics.GetOrRegisterCounter("pusher.subscribe-push.chunks-in-batch", nil).Inc(int64(chunksInBatch))
+				}
+				chunksInBatch = 0
+				batchStartTime = time.Now()
+
+				// and start iterating on Push index from the beginning
+				chunks, unsubscribe = p.store.SubscribePush(ctx)
+				// reset timer to go off after retryInterval
+				timer.Reset(retryInterval)
+			}()
+
+		case <-p.quit:
+			if unsubscribe != nil {
+				unsubscribe()
+			}
+			return
+		}
+	}
+}
+
+// handleReceiptMsg is a handler for pssReceiptTopic that
+// - deserialises receiptMsg and
+// - sends the receipted address on a channel
+func (p *Pusher) handleReceiptMsg(msg []byte) error {
+	receipt, err := decodeReceiptMsg(msg)
+	if err != nil {
+		return err
+	}
+	p.logger.Trace("handleReceiptMsg", "receipt", hex.EncodeToString(receipt.Addr))
+	go p.pushReceipt(receipt.Addr)
```

Comment:
with async pss this is not needed i think

Reply:
We had worse results without it, even if recent pss changes suggest that it is not needed. We did not get much deeper why is that the case.